mips/intr.c

```diff
@@ -275,27 +301,41 @@ void mips_exc_handler(exc_frame_t *frame) {
   if (!handler)
     kernel_oops(frame);
 
-  /* Enable interrupts only if you're about to handle an exception that came
-   * from a context that had hardware interrupts enabled.
-   * We don't want to enable interrupts if an exception was generated
-   * in a critical section running with interrupts disabled. */
-  if ((code != EXC_INTR) && (frame->sr & SR_IE))
-    intr_enable();
-
-  (*handler)(frame);
-
-  /* From now on till the end of this procedure interrupts are enabled. */
-  if (code == EXC_INTR)
-    intr_enable();
+  if (!user_mode && (!(frame->sr & SR_IE) || preempt_disabled())) {
+    /* Case 2b & 2c: we came from kernel-space, interrupts or preemption were
+     * disabled, so switching out is forbidden! */
```

Comment:
If this is case 2b, we should enable interrupts if this is exception handling. Currently case 2b is handled always with interrupts disabled.

Reply:
Right... I'll have to give it another round of thinking.

Reply:
I'll keep it as is. I added somewhat lengthy explanation of reasoning behind it.